File: src/main/java/org/moinex/service/I18nService.java
package org.moinex.service;

import java.util.List;
import java.util.Locale;
import java.util.MissingResourceException;
import java.util.ResourceBundle;
import java.util.prefs.Preferences;
import org.springframework.stereotype.Service;

@Service
public class I18nService {
    private static final String BUNDLE_BASE_NAME = "i18n.messages";
    private static final String PREF_KEY_LOCALE = "ui.locale";

    private final Preferences preferences = Preferences.userRoot().node("org.moinex");

    private Locale locale;

    public I18nService() {
        this.locale = resolveInitialLocale();
    }

    public Locale getLocale() {
        return locale;
    }

    public void setLocale(Locale locale) {
        if (locale == null) {
            return;
        }

        this.locale = locale;
        preferences.put(PREF_KEY_LOCALE, toPreferenceValue(locale));
    }

    public List<Locale> getSupportedLocales() {
        return List.of(Locale.forLanguageTag("pt-BR"), Locale.ENGLISH);
    }

    public ResourceBundle getBundle() {
        try {
            return ResourceBundle.getBundle(BUNDLE_BASE_NAME, locale);
        } catch (MissingResourceException ex) {
            return ResourceBundle.getBundle(BUNDLE_BASE_NAME, Locale.ENGLISH);
        }
    }

    public String tr(String key) {
        if (key == null) {
            return "";
        }

        try {
            return getBundle().getString(key);
        } catch (MissingResourceException ex) {
            return key;
        }
    }

    private Locale resolveInitialLocale() {
        String stored = preferences.get(PREF_KEY_LOCALE, null);
        if (stored != null && !stored.isBlank()) {
            return Locale.forLanguageTag(stored.replace('_', '-'));
        }

        Locale system = Locale.getDefault();
        if (system != null) {
            for (Locale supported : getSupportedLocales()) {
                if (supported.getLanguage().equalsIgnoreCase(system.getLanguage())) {
                    if (supported.getCountry().isEmpty()
                            || supported.getCountry().equalsIgnoreCase(system.getCountry())) {
                        return supported;
                    }
                }
            }
        }

        return Locale.forLanguageTag("pt-BR");
    }

    private String toPreferenceValue(Locale locale) {
        if (locale.getCountry().isBlank()) {
            return locale.getLanguage();
        }
        return locale.getLanguage() + "_" + locale.getCountry();
    }
}

File: src/main/kotlin/org/moinex/service/I18nService.kt
/*
 * Filename: I18nService.kt (original filename: I18nService.java)
 * Created on: December 29, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 03/08/2026
 */

package org.moinex.service

import org.springframework.stereotype.Service
import java.util.Locale
import java.util.MissingResourceException
import java.util.ResourceBundle
import java.util.prefs.Preferences

@Service
class I18nService {
    companion object {
        const val BUNDLE_BASE_NAME = "i18n.messages"
        const val PREF_KEY_LOCALE = "ui.locale"
        const val BRAZILIAN_PORTUGUESE_TAG = "pt-BR"
        val SUPPORTED_LOCALES = listOf(Locale.forLanguageTag(BRAZILIAN_PORTUGUESE_TAG), Locale.ENGLISH)
        val preferences: Preferences = Preferences.userRoot().node("org.moinex")
    }

    var locale: Locale = resolveInitialLocale()
        set(value) {
            field = value
            preferences.put(PREF_KEY_LOCALE, toPreferenceValue(value))
            bundleCache = null
        }

    private var bundleCache: ResourceBundle? = null

    fun getSupportedLocales(): List<Locale> = SUPPORTED_LOCALES

    fun getBundle(): ResourceBundle =
        bundleCache ?: try {
            ResourceBundle.getBundle(BUNDLE_BASE_NAME, locale).also { bundleCache = it }
        } catch (_: MissingResourceException) {
            ResourceBundle.getBundle(BUNDLE_BASE_NAME, Locale.ENGLISH).also { bundleCache = it }
        }

    fun tr(key: String?): String =
        key?.let {
            try {
                getBundle().getString(it)
            } catch (_: MissingResourceException) {
                it
            }
        } ?: ""

    private fun resolveInitialLocale(): Locale {
        val stored =
            preferences[PREF_KEY_LOCALE, null]
                ?.takeIf { it.isNotBlank() }
                ?.let { Locale.forLanguageTag(it.replace('_', '-')) }

        if (stored != null) return stored

        val system = Locale.getDefault()
        return system?.let { systemLocale ->
            SUPPORTED_LOCALES.firstOrNull { supported ->
                supported.language.equals(systemLocale.language, ignoreCase = true) &&
                    (
                        supported.country.isEmpty() ||
                            supported.country.equals(systemLocale.country, ignoreCase = true)
                    )
            }
        } ?: Locale.forLanguageTag(BRAZILIAN_PORTUGUESE_TAG)
    }

    private fun toPreferenceValue(locale: Locale): String =
        locale.language +
            locale.country
                .takeIf { it.isNotBlank() }
                ?.let { "_$it" }
                .orEmpty()
}
